README.md
# Born Again Netscape's JavaScript Entities (entities.js)

**JavaScript Entities** was a feature supported in Netscape that allowed to
use JavaScript within HTML attribute values, where a string enclosed within
`&{` and `};` was parsed and evaluated as valid JavaScript code.

For example,

```
<script type="text/javascript">
	var myCSS = "color: red;";
</script>

<p style="&{myCSS};">Hi!</p>
```

was evaluated as

```html
<p style="color: red;">Hi!</p>
```

This framework allows to use again Netscape's feature together with modern
HTML5. All you need to do is to include `entities.js` in your document,
then let it parse the page:

```html
<script type="text/javascript" src="entities-min.js"></script>
<script type="text/javascript">
addEventListener("DOMContentLoaded", function () {
	JSEntities.parseTree(document.documentElement);
}, false);
</script>
```

## Ambiguous ampersands?

If you wish to safely write an ampersand in HTML (as in, for example,
&ldquo;R&amp;B&rdquo;) you might want to encode it (`&amp;`), since
non-encoded ampersands (`&`) could combine with the text following and result
in *ambiguous ampersands* &ndash; i.e., invalid HTML code. However, **not
all the unencoded ampersands are ambiguous ampersands.** According to the
[WHATWG HTML Living Standard][1],

> An **ambiguous ampersand** is a `U+0026 AMPERSAND` character (&amp;)
  that is followed by one or more [ASCII alphanumerics][2], followed by a
  `U+003B SEMICOLON character` (;), where these characters do not match any
  of the names given in the [named character references][3] section.

The leading sequence of JavaScript entities is `&{` &ndash; which is an
ampersand character *not* followed by any ASCII alphanumeric (but by a curly
bracket instead). Hence, **JavaScript entities are perfectly valid HTML code**.

Here follows an example table that shows a few cases of ambiguous and
non-ambiguous ampersands:

| Sequence	   | Is it ambiguous? |
|------------------|------------------|
| `&foo;`	   | Yes	      |
| `&area51;`	   | Yes	      |
| `&foo`	   | No		      |
| `&{foo};`	   | No		      |
| `&!foo;`	   | No		      |
| `&{;`		   | No		      |
| `&{foo;`	   | No		      |
| `&{};`	   | No		      |

## Notes

* The [HTML 4.01 specification][4] reserves a syntax for the &ldquo;future
  support of script macros&rdquo; in HTML attributes (i.e., JavaScript
  entities). Sadly this syntax has never been incorporated into later
  standards, and currently &ldquo;script macros&rdquo; are not supported by
  any browser. This library can be considered as a general &ldquo;script macros
  polyfill&rdquo; for all browsers that do not natively support this feature.
* The function `JSEntities.parseTree()` replaces the text of *all* the
  attributes contained in a DOM tree. Since some of these attributes might
  have been assigned via JavaScript and might no longer contain strings
  (as, for example, `mySpanElement.onlick = clickMe;` &ndash; where `typeof
  mySpanElement.onlick` is usually a `"function"` and not a `"string"`),
  it is preferable to launch `JSEntities.parseTree()` *before any other script*.
* The string segments enclosed within `&{` and `};` are passed **verbatim**
  to [`eval()`][5] and executed in the global scope.
* A right curly bracket followed by a semicolon (`};`) *does not* end the
  Javascript entity unless it no longer expresses a syntactically valid
  JavaScript symbol &ndash; as it is in the segment `'red' };` in `<p
  style="color: &{var myObject = { 'color': 'red' }; myObject.color;};">Hi!</p>`
* Within node attributes the characters `<`, `>` and `"` must *always* be
  encoded (respectively, `&lt;`, `&gt;` and `&quot;`). As for the character `&`,
  it must be encoded (`&amp;`) when it represents an ambiguous ampersand &ndash;
  as in, for example, `var bContinue=bAccept&&bAvailable;` (see above). Instead,
  within JavaScript strings passed directly to `JSEntity.parseString()`,
  JavaScript entities are the only entities supported.

Enjoy the entities!

[1]: https://html.spec.whatwg.org/multipage/syntax.html#syntax-ambiguous-ampersand
[2]: https://infra.spec.whatwg.org/#ascii-alphanumeric
[3]: https://html.spec.whatwg.org/multipage/syntax.html#named-character-references
[4]: http://www.w3.org/TR/REC-html40/appendix/notes.html#h-B.7.1
[5]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval



ChangeLog.md
# Change Log

## 1.0.4 -- 2017-10-20

Changes:

* Code review
* Documentation

## 1.0.3 -- 2017-03-18

Changes:

* Documentation

## 1.0.2 -- 2017-03-04

Changes:

* Code review
* Documentation

## 1.0.1 -- 2017-03-01

Changes:

* Object `Entities` renamed `JSEntities`
* Documentation

## 1.0.0 -- 2017-02-27

JSEntities 1.0.0 is released!


entities-min.js
/**
	@source		https://github.com/madmurphy/entities.js/
	@author		madmurphy
	@copyright 	GNU Public License v3
	@version	1.0.4
**/
"use strict";var JSEntities={parseString:function(s){for(var t,i=0,r=s.length,e=[],a=0,n=0,u=0,h=1;r>n;n++)if(t=s.charCodeAt(n),h=38!==t||1985&(1^h)?123===t&&5&(1^h)&&!(1986&h)?4035&h|8:125!==t||1987&h?39!==t||1955&h?34!==t||1891&h?(10===t||13===t)&&512&h?193&h:92===t?4035&h^2:47!==t||193&h?42===t?256&h?2529&h|2048:32&h?449&h|256:1024&h?4033&h:1985&h:32&h?3265&h|3072:1024&h?4033&h:1985&h:1824&(32^h)?2050&(2048^h)?1792&h?4033&h:225&h|32:193&h|h<<1&2048:705&h|512:4035&h^128:4035&h^64:4051&h|16:4037&h|4,8&h)1&h&&(a=n,h&=4094),u++;else if(16&h&&0===--u){if(r>n+1&&59===s.charCodeAt(n+1)){e.push(s.substring(i,a-1));try{e.push(window.eval(s.substring(a+1,n)))}catch(o){console.log("JSEntities, parsing error - "+o.message+" [skip]"),e.push(s.substring(a-1,n+2))}i=n+2,n++}h=1}return r>i&&e.push(s.slice(i)),e.join("")},parseTree:function(s){var t;if(s.hasAttributes&&s.hasAttributes())for(var i=0;i<s.attributes.length;i++)t=s.attributes[i],t.value=this.parseString(t.value);if(s.hasChildNodes())for(t=s.firstChild;t;this.parseTree(t),t=t.nextSibling);}};


entities_min.js
/**
	@source		https://github.com/madmurphy/entities.js/
	@author		madmurphy
	@copyright 	GNU Public License v3
	@version	1.0.3
**/
"use strict";var JSEntities={parseString:function(a){for(var b,c=0,d=0,e=0,f=a.length,g="",h=1,i=0;i<f;i++)if(b=a.charCodeAt(i),h=38!==b||1987&(1^h)?123===b&&5&(1^h)&&!(1986&h)?4035&h|8:125!==b||1987&h?39!==b||1955&h?34!==b||1891&h?47!==b||193&h?92===b?4035&h^2:42===b?256&h?2529&h|2048:32&h?449&h|256:1024&h?4033&h:1985&h:(10===b||13===b)&&512&h?193&h:32&h?3265&h|3072:1024&h?4033&h:1985&h:1824&(32^h)?2050&(2048^h)?1792&h?4033&h:225&h|32:193&h|h<<1&2048:705&h|512:4035&h^128:4035&h^64:4051&h|16:4039&h|4,8&h)1&h&&(d=i,h&=4094),e++;else if(16&h&&0===--e){if(i+1<f&&59===a.charCodeAt(i+1)){g+=a.substring(c,d-1);try{g+=window.eval(a.substring(d+1,i))}catch(b){console.log("JSEntities, parsing error - "+b.message+" [skip]"),g+=a.substring(d-1,i+2)}c=i+2,i++}h=1}return c<f&&(g+=a.slice(c)),g},parseTree:function(a){var b,c;if(a.hasAttributes&&a.hasAttributes())for(b=0;b<a.attributes.length;b++)c=a.attributes[b],c.value=this.parseString(c.value);if(a.hasChildNodes())for(c=a.firstChild;c;this.parseTree(c),c=c.nextSibling);}};


entities.js
/*\
|*|
|*|	:: entities.js ::
|*|
|*|	Born Again Netscape's JavaScript Entities
|*|
|*|	Version 1.0.4
|*|
|*|	https://github.com/madmurphy/entities.js/
|*|
|*|	JS Syntax:
|*|
|*|		* `JSEntities.parseTree(document.documentElement)`
|*|		* `console.log(JSEntities.parseString("Today is &{new Date();};"));`
|*|
|*|	HTML Example:
|*|
|*|		* `<p title="&{new Date();};" style="color: &{(new Date()).getHours() & 1 ? 'red' : 'blue';};">Hello world!</p>`
|*|
|*|	This framework is released under the GNU Public License, version 3 or later.
|*|	http://www.gnu.org/licenses/gpl-3.0-standalone.html
|*|
\*/

"use strict";

var JSEntities = {

	/**

	@brief		Parses and evaluates all the JavaScript entities nested within a string
	@param		sInput		The string to parse and evaluate
	@return		The evaluated string

	**/
	"parseString": function (sInput) {

		var nIdxA = 0, nLen = sInput.length, aParts = [];

		/*

			Mask `nAbc` (12 bits used):

			FLAG_1		We are *not* inside a JavaScript entity
			FLAG_2		We are in an odd sequence of backslashes
			FLAG_4		Ampersand found AND `FLAG_1 === true`
			FLAG_8		Left curly bracket found inside or at the
					beginning of a JavaScript entity
			FLAG_16		Right curly bracket out of quote found inside or
					at the end of a JavaScript entity
			FLAG_32		Leading slash found
			FLAG_64		We are inside a single-quote string expression
			FLAG_128	We are inside a double-quote string expression
			FLAG_256	We are inside a multiline comment
			FLAG_512	We are inside an inline comment
			FLAG_1024	We are inside a regular expression
			FLAG_2048	Next slash will close a slash expression (either
					a regular expression or a multiline comment)

		*/

		for (var /* let */ nChr, nIdxB = 0, nIdxC = 0, nCurlies = 0, nAbc = 1; nIdxC < nLen; nIdxC++) {

			nChr = sInput.charCodeAt(nIdxC);

			nAbc	=	nChr === 38 && !((nAbc ^ 1) & 1985) ?				/* `&` */
						(nAbc & 4037) | 4
					: nChr === 123 && ((nAbc ^ 1) & 5) && !(nAbc & 1986) ?		/* `{` */
						(nAbc & 4035) | 8
					: nChr === 125 && !(nAbc & 1987) ?				/* `}` */
						(nAbc & 4051) | 16
					: nChr === 39 && !(nAbc & 1955) ?				/* `'` */
						(nAbc & 4035) ^ 64
					: nChr === 34 && !(nAbc & 1891) ?				/* `"` */
						(nAbc & 4035) ^ 128
					: (nChr === 10 || nChr === 13) && (nAbc & 512) ?		/* `\n` or `\r` */
						nAbc & 193
					: nChr === 92 ?							/* `\` */
						(nAbc & 4035) ^ 2
					: nChr === 47 && !(nAbc & 193) ?				/* `/` */
						(
							!((nAbc ^ 32) & 1824) ?
								(nAbc & 705) | 512
							: !((nAbc ^ 2048) & 2050) ?
								(nAbc & 193) | (nAbc << 1 & 2048)
							: nAbc & 1792 ?
								nAbc & 4033
							:
								(nAbc & 225) | 32
						)
					: nChr === 42 ?							/* `*` */
						(
							nAbc & 256 ?
								(nAbc & 2529) | 2048
							: nAbc & 32 ?
								(nAbc & 449) | 256
							: nAbc & 1024 ?
								nAbc & 4033
							:
								nAbc & 1985
						)
					: nAbc & 32 ?
						(nAbc & 3265) | 3072
					: nAbc & 1024 ?
						nAbc & 4033
					:
						nAbc & 1985;

			if (nAbc & 8) {

				if (nAbc & 1) {

					nIdxB = nIdxC;
					nAbc &= 4094;

				}

				nCurlies++;

			} else if ((nAbc & 16) && --nCurlies === 0) {

				if (nIdxC + 1 < nLen && sInput.charCodeAt(nIdxC + 1) === 59 /* `;` */) {

					aParts.push(sInput.substring(nIdxA, nIdxB - 1));

					try {

						aParts.push(window.eval(sInput.substring(nIdxB + 1, nIdxC)));

					} catch (oErr) {

						console.log("JSEntities, parsing error - " + oErr.message + " [skip]");
						aParts.push(sInput.substring(nIdxB - 1, nIdxC + 2));

					}

					nIdxA = nIdxC + 2;
					nIdxC++;

				}

				nAbc = 1;

			}

		}

		if (nIdxA < nLen) {

			aParts.push(sInput.slice(nIdxA));

		}

		return aParts.join("");

	},

	/**

	@brief		Parses all the attributes contained in a DOM tree
	@param		oParent		The DOM tree to parse
	@return		Nothing

	**/
	"parseTree": function (oParent) {

		var oIter;

		if (oParent.hasAttributes && oParent.hasAttributes()) {

			for (var /* let */ nIdx = 0; nIdx < oParent.attributes.length; nIdx++) {

				oIter = oParent.attributes[nIdx];
				oIter.value = this.parseString(oIter.value);

			}

		}

		if (oParent.hasChildNodes()) {

			for (oIter = oParent.firstChild; oIter; this.parseTree(oIter), oIter = oIter.nextSibling);

		}

	}

};

